lib/spack/spack/cmd/checksum.py

```diff
@@ -56,33 +56,34 @@ def checksum(parser, args):
     # Get the package we're going to generate checksums for
     pkg = spack.repo.get(args.package)
 
+    # Otherwise, see what versions we can find online
     url_dict = {}
-    if args.versions:
-        # If the user asked for specific versions, use those
-        for version in args.versions:
+    versions = args.versions
+    if (not versions) and args.preferred:
+        versions = [preferred_version(pkg)]
+
+    if versions:
+        remote_versions = None
+        for version in versions:
             version = ver(version)
             if not isinstance(version, Version):
                 tty.die("Cannot generate checksums for version lists or "
                         "version ranges. Use unambiguous versions.")
-            url_dict[version] = pkg.url_for_version(version)
-    elif args.preferred:
-        version = preferred_version(pkg)
-        url_dict = dict([(version, pkg.url_for_version(version))])
+            url = pkg.find_valid_url_for_version(version)
+            if url is not None:
+                url_dict[version] = url
+                continue
+            # if we get here, it's because no valid url was provided by the package
+            # do expensive fallback to try to recover
+            if remote_versions is None:
+                remote_versions = pkg.fetch_remote_versions()
+            if version in remote_versions:
+                url_dict[version] = remote_versions[version]
```

Comment:
(question) this looks like a preexisting problem but to make sure: if this `if` check fails, it looks like we should raise an exception here: at this point there is no URL available to fetch the version and checksum it.

Reply:
Right, the current behavior is it only fails if none of the versions are found, rather than if any of them are not found.